File: python/rank_tracker.py
#!/usr/bin/env python3
"""Track where an app ranks for a keyword (ASO rank tracking) with StorePulse.

Search the store for a keyword and report the position of a specific app — the
core of any ASO / keyword rank-tracking workflow. Run it on a schedule and log
the rank over time.

StorePulse — Google Play app metadata, reviews, search rankings & charts as a
clean, cached JSON API. Stop maintaining your own Play scraper:
https://rapidapi.com/egosumkira/api/storepulse

Usage:
    export RAPIDAPI_KEY=...
    python rank_tracker.py "music streaming" com.spotify.music [country] [lang]
"""
import os
import sys
from datetime import datetime, timezone

import requests

HOST = "storepulse.p.rapidapi.com"


def find_rank(query, target_app_id, country="us", lang="en"):
    key = os.environ["RAPIDAPI_KEY"]
    resp = requests.get(
        f"https://{HOST}/v1/search",
        params={"q": query, "country": country, "lang": lang},
        headers={"X-RapidAPI-Key": key, "X-RapidAPI-Host": HOST},
        timeout=30,
    )
    resp.raise_for_status()
    for r in resp.json()["results"]:
        if r["app_id"] == target_app_id:
            return r["rank"]
    return None  # not in the returned page of results


def main():
    if len(sys.argv) < 3:
        sys.exit('usage: python rank_tracker.py "<query>" <package_id> [country] [lang]')
    query = sys.argv[1]
    target = sys.argv[2]
    country = sys.argv[3] if len(sys.argv) > 3 else "us"
    lang = sys.argv[4] if len(sys.argv) > 4 else "en"

    rank = find_rank(query, target, country, lang)
    now = datetime.now(timezone.utc).isoformat()
    if rank is None:
        print(f"{now}  {target}  query={query!r}  country={country}  rank=NOT_FOUND")
    else:
        print(f"{now}  {target}  query={query!r}  country={country}  rank={rank}")


if __name__ == "__main__":
    main()

File: node/rank_tracker.js
#!/usr/bin/env node
// Track where an app ranks for a keyword (ASO rank tracking) with StorePulse.
//
// Search the store for a keyword and report the position of a specific app — the
// core of any ASO / keyword rank-tracking workflow. Run it on a schedule and log
// the rank over time.
//
// StorePulse — Google Play app metadata, reviews, search rankings & charts as a
// clean, cached JSON API. Stop maintaining your own Play scraper:
// https://rapidapi.com/egosumkira/api/storepulse
//
// Usage:
//   export RAPIDAPI_KEY=...
//   node rank_tracker.js "music streaming" com.spotify.music [country] [lang]
//
// Requires Node 18+ (built-in fetch).

const HOST = "storepulse.p.rapidapi.com";

async function findRank(query, targetAppId, country = "us", lang = "en") {
  const key = process.env.RAPIDAPI_KEY;
  if (!key) throw new Error("set RAPIDAPI_KEY in your environment");

  const url = new URL(`https://${HOST}/v1/search`);
  url.search = new URLSearchParams({ q: query, country, lang }).toString();

  const resp = await fetch(url, {
    headers: { "X-RapidAPI-Key": key, "X-RapidAPI-Host": HOST },
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  const data = await resp.json();
  const hit = data.results.find((r) => r.app_id === targetAppId);
  return hit ? hit.rank : null; // null = not in the returned page of results
}

async function main() {
  const [query, target, country = "us", lang = "en"] = process.argv.slice(2);
  if (!query || !target) {
    console.error('usage: node rank_tracker.js "<query>" <package_id> [country] [lang]');
    process.exit(1);
  }

  const rank = await findRank(query, target, country, lang);
  const now = new Date().toISOString();
  const value = rank === null ? "NOT_FOUND" : rank;
  console.log(`${now}  ${target}  query="${query}"  country=${country}  rank=${value}`);
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});
